httpcore5-h2/src/main/java/org/apache/hc/core5/http2/impl/nio/H2OnlyClientProtocolNegotiator.java

```diff
@@ -104,7 +106,8 @@ private void writeOutPreface(final IOSession session) throws IOException  {
         }
         if (!preface.hasRemaining()) {
             session.clearEvent(SelectionKey.OP_WRITE);
-            startProtocol(HttpVersion.HTTP_2, new ClientH2IOEventHandler(http2StreamHandlerFactory.create(ioSession)), null);
+            final ByteBuffer data = inBuf != null ? inBuf.data() : null;
```

Comment:
We can avoid the unnecessary additional volatile read (I find this pattern helpful in general, especially when null checks are involved, though I realize in this case we never transition from non-null to null)
```suggestion
            final BufferedData inBufSnapshot = inBuf;
            final ByteBuffer data = inBufSnapshot != null ? inBufSnapshot.data() : null;
```

Reply:
@carterkozak `IOEventHandler`s can be called by a single I/O event dispatch thread at a time by design and therefore do not need to synchronize access to their internal structures if they do not delegate event handling to external handlers.